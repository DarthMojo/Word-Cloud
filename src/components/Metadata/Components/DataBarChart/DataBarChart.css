.ChartTitle{
    display: flex;
    font-size: 2em;
    justify-content: center;
    text-decoration: underline;
}
