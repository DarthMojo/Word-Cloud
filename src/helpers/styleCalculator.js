/**
 * 1. Each topic should have one of 6 different text sizes,
 * with the most popular topics largest, and least popular smallest
 *
 * 2.A topic with a sentiment score > 60 should be displayed in green
 *
 * 3.A topic with a sentiment score < 40 should be displayed in red
 *
 * 4.Other topics should be displayed in grey
 */
import type Dimension from "../config/types";
import { sentimentScoreMin,
    sentimentScoreMax,
    defaultColor,
    size } from "../config/static";


/**
 *
 * @type {{getTopicSize: (function(*, *)),
 * getTopicColor: (function(*)),
 * windowSize: (function(*=)),
 * rotation: (function(): number)}}
 */
const styleCalculator = {
    /**
     *
     * @param wordPopularity
     * @param popularityRange
     * @return {number}
     */
    getTopicSize: (wordPopularity, popularityRange): number => {
        const unitRange = ((popularityRange.max - popularityRange.min) / size.max) || 0;
        const sizeValue: number = Array.from({ length: size.max }, (v, k): number => k + 1)
            .find((sizeIndex: number): number =>
                wordPopularity <= popularityRange.min + (sizeIndex * unitRange));
        return Math.round((sizeValue / size.max) * 100);
    },
    /**
     *
     * @param sentimentScore
     * @return String
     */
    getTopicColor: (sentimentScore): string => {
        if (sentimentScore > sentimentScoreMax.score) {
            return sentimentScoreMax.color;
        } else if (sentimentScore < sentimentScoreMin.score) {
            return sentimentScoreMin.color;
        }
        return defaultColor;
    },
    /**
     *
     * @param factor
     * @return {{width: number, height: number}}
     */
    windowSize: (factor = 1): Dimension => {
        return {
            width: Math.round(window.innerWidth * factor),
            height: Math.round(window.innerHeight * factor)
        };
    },
    /**
     *@return {number}
     */
    rotation: (): number => (~~(Math.random() * size.max) - size.max * 0.5) * 30
};
export default styleCalculator;
