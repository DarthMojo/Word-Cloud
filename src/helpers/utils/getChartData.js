/**
 *
 * @param values
 * @return {Array}
 */
const getChartData = (values) => {
    const data = [];
    for (const [key, value] of Object.entries(values)) {
        data.push({ x: key, y: value });
    }
    return data;
};
export default getChartData;
