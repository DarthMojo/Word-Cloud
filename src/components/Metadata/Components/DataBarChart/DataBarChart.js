import React from "react";
import type { Element } from "react";
import { BarChart } from "react-easy-chart";
import { minWindowWidth, chartSizeFactor, chartMargin } from "../../../../config/static";

import "./DataBarChart.css";

type DataBarChartType = {
    data: Array<Object>;
    dimensions: Object;
    chartTitle: string
};

/**
 *
 * @param data
 * @param dimensions
 * @param chartTitle
 * @return {Element}
 * @constructor
 */
const DataBarChart = ({ data, dimensions, chartTitle }: DataBarChartType): Element<any> => {
    return (<div>
        <div className={"ChartTitle"}>{chartTitle}</div>
        <BarChart
            axes={(dimensions.width) > minWindowWidth}
            colorBars
            grid
            height={dimensions.height * chartSizeFactor}
            width={dimensions.width}
            data={data}
            yTickNumber={5}
            margin={chartMargin}
        /></div>);
};
export default DataBarChart;
