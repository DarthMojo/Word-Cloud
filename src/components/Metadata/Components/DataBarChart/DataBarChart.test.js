/* eslint-disable no-undef */
import React from "react";
import styleCalculator from "../../../../helpers/styleCalculator";
import DataBarChart from "./DataBarChart";
import { shallowToJson } from "enzyme-to-json";
import getChartData from "../../../../helpers/utils/getChartData";
import data from "../../../../config/data/topics.json";

test("render DataBarChart()", () => {
    const dimensions = styleCalculator.windowSize();

    const wordData = data.topics.find((wordItem): Object<*> =>
        wordItem.id === "1751295897__Berlin");

    const wrapper = shallow(
        <DataBarChart data={getChartData(wordData.pageType)}
            dimensions={dimensions} label={{ x: "Channel", y: "Hits" }}/>
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
});
