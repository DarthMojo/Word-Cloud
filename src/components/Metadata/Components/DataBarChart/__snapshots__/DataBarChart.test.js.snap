// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render DataBarChart() 1`] = `
<div>
  <div
    className="ChartTitle"
  />
  <BarChart
    axes={true}
    axisLabels={
      Object {
        "x": "",
        "y": "",
      }
    }
    barWidth={10}
    clickHandler={[Function]}
    colorBars={true}
    data={
      Array [
        Object {
          "x": "blog",
          "y": 17,
        },
        Object {
          "x": "facebook",
          "y": 56,
        },
        Object {
          "x": "forum",
          "y": 22,
        },
        Object {
          "x": "general",
          "y": 5,
        },
        Object {
          "x": "image",
          "y": 0,
        },
        Object {
          "x": "news",
          "y": 26,
        },
        Object {
          "x": "review",
          "y": 1,
        },
        Object {
          "x": "twitter",
          "y": 35,
        },
        Object {
          "x": "video",
          "y": 3,
        },
      ]
    }
    datePattern="%d-%b-%y"
    grid={true}
    height={576}
    interpolate="linear"
    lineData={Array []}
    margin={
      Object {
        "bottom": 50,
        "left": 50,
        "right": 50,
        "top": 50,
      }
    }
    mouseMoveHandler={[Function]}
    mouseOutHandler={[Function]}
    mouseOverHandler={[Function]}
    width={1024}
    xType="text"
    y2Type="linear"
    yTickNumber={5}
    yType="linear"
  />
</div>
`;
